### ISSUE TITLE
UnionGraphType.ResolveType delegate requires an instance to be returned

### ISSUE THREAD

**Issue by Shane32** [graphql-dotnet/graphql-dotnet#2684]
The `UnionGraphType.ResolveType` delegate requires that the implementer return an instance of a graph type rather than a `Type`.  With the typical way of configuring a schema with types rather than instances (when using DI), this can break schema initialization when used with the recommended pattern of transient registration of graph types within the DI provider (with the expectation that each graph type is instianted only once by `SchemaTypes`).  What happens is that because an instance is required to be returned, the instance must be created manually or by DI in the constructor of the union type.  Any other graph that uses the same graph type can cause the schema to initialize only one of the two instances, leaving an uninitialized instance within the schema.  Graph resolution then fails when the uninitialized instance is used by the execution strategy.

In a similar vein, any time that the `ResolveType` delegate is used, it must return the exact same instance of the selected graph type as is used elsewhere in the graph.  This can be an issue for dynamically generated graph structures even without DI.

Workarounds:
- Register any types needed by `ResolveType` delegates with a scoped or singleton lifetime, and pull those graph types in via dependency injection from the union graph type's constructor; or
- Use the `IsTypeOf` delegate rather than the `ResolveType` delegate.

Possible fixes:
- Add a delegate similar to the `ResolveType` delegate which returns a type rather than an instance, which is then resolved internally to the correctly initialized instance.

See issue:
- #2674

**Comment by Shane32** [comment]
This is labeled a bug as it can cause problems when using DI and union types as described in the documentation.

**Comment by Shane32** [comment]
I believe that @joemcbride has indicated in another issue (not sure which at the moment) that `new GraphQLTypeReference` is the suggested workaround at this time when using the `ResolveType` delegate.  See:
- #775
